### PR TITLE
fix(security): harden dependabot internal scope updates

### LINKED ISSUE #1803: Security: harden dependency update workflow against supply-chain confusion
## Description
Implement an LLM-friendly, independently shippable backlog item for **harden dependency update workflow against supply-chain confusion**. This issue should stay narrow enough for one agent/PR while producing durable behavior, tests, and documentation where appropriate.

## Affected locations
- Input parsing and untrusted text handling
- Web/API/control-plane boundaries
- Approval, token, and state mutation paths

## Recommendation
Add the smallest safe control that addresses this risk, prefer deny-by-default behavior for unsafe paths, document operator override semantics, and include regression tests for both allowed and denied flows.

## Goals
Deliver harden dependency update workflow against supply-chain confusion as a discrete capability or hardening change.
Make the behavior deterministic enough for autonomous workers to verify.
Keep the implementation small enough for a one-issue worker and one PR.
Capture operator/user-facing behavior in docs or help text when applicable.

## Expected outcomes
harden dependency update workflow against supply-chain confusion is available, documented, and test-backed.
Future PM/worker handoffs can point at this issue without needing extra interpretation.
Failure modes are explicit and produce actionable messages instead of silent drift.
The change does not broaden into adjacent backlog work or unrelated refactors.

## Suggested implementation slices
Inventory the current code/docs paths that own this behavior.
Add or update the narrow model/config/schema needed for the feature or guardrail.
Implement the smallest runtime/docs change that satisfies the acceptance criteria.
Add focused fixtures/tests for success, failure, and edge cases.
Update docs/help output and handoff notes so later agents can use it.

## Acceptance criteria
- [ ] There is a dedicated implementation or documentation change for harden dependency update workflow against supply-chain confusion.
- [ ] The control defaults to the safest practical behavior and requires explicit override for risky cases.
- [ ] At least one automated test, fixture, or deterministic verifier covers the intended behavior.
- [ ] Sensitive payloads/secrets are not printed in logs, test output, issue comments, or telemetry.
- [ ] At least one negative/edge case is covered so the failure mode is explicit.
- [ ] Docs/help text or inline operator guidance explains how to use or interpret the behavior.
- [ ] The PR description links this issue and summarizes verification evidence.
- [ ] No unrelated production behavior is changed outside the stated scope.

## LLM handoff notes
- Treat this as a one-issue/one-PR unit.
- Start by searching for existing implementations and tests before adding new surfaces.
- Prefer structured outputs and fixtures over prose-only behavior.
- If this overlaps an existing open issue, comment with evidence and stop for PM deduplication instead of creating an unscoped implementation.
- Do not perform broad refactors or change unrelated worker/PM behavior while implementing this ticket.

## Source
Created by the 2026-07-11 fourth strategic backlog expansion run requesting 100 issues. Topic: **Vulnerabilities**.

### PR DESCRIPTION
**Pull request by djm204**
## Summary
- Add a fail-closed Dependabot supply-chain guard that requires npm update entries to ignore `@franken/*` for major/minor/patch registry updates.
- Exclude `@franken/*` from the broad npm Dependabot group so internal workspace package versions remain release-workflow controlled.
- Wire the guard into CI and document the operator workflow in security/quickstart docs.

## Verification
- `node scripts/check-dependabot-supply-chain.mjs`
- `npm run test:root -- tests/unit/dependency-ci-guards.test.ts`
- `npm run check:dependabot-supply-chain`
- `node scripts/check-workspace-deps.mjs`
- `npm run lint:security`
- `npm run typecheck`
- `npm run lint` (passes with existing warnings)
- `npm run build`
- `npm run audit:security`

Closes #1803

### CODE REVIEW

**Comment by djm204** [comment]
@codex review

File: scripts/check-dependabot-supply-chain.mjs [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+      continue;
+    }
+    if (/^\s{4}ignore:\s*$/.test(line)) {
+      mode = 'ignore';
+      groupName = null;
+      continue;
+    }
+
+    const groupMatch = line.match(/^ {6}([^\s][^:]*):\s*$/);
+    if (groupMatch && mode === 'groups') {
```
[the remark this answers is not in the document]
Comment (djm204) [reply scripts/check-dependabot-supply-chain.mjs]: Addressed in bfc5156b: switched the guard to parse Dependabot YAML with js-yaml so every group and inline list form is recognized, require @franken/* exclusions on every npm group (including dependency-type groups), and require the @franken/* ignore rule without update-types filters so security updates are blocked too. Added regression coverage for these cases.

File: scripts/check-dependabot-supply-chain.mjs [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+      groupName = unquote(groupMatch[1]);
+      update.groups[groupName] = update.groups[groupName] ?? {};
+      continue;
+    }
+    if (line.match(/^\s{8}patterns:\s*$/) && mode === 'groups' && groupName) {
```
[the remark this answers is not in the document]
Comment (djm204) [reply scripts/check-dependabot-supply-chain.mjs]: Addressed in bfc5156b: switched the guard to parse Dependabot YAML with js-yaml so every group and inline list form is recognized, require @franken/* exclusions on every npm group (including dependency-type groups), and require the @franken/* ignore rule without update-types filters so security updates are blocked too. Added regression coverage for these cases.

File: .github/dependabot.yml [hunk 3]
```diff
@@ -9,6 +9,14 @@ updates:
       npm-security-and-maintenance:
         patterns:
           - "*"
+        exclude-patterns:
+          - "@franken/*"
+    ignore:
+      - dependency-name: "@franken/*"
+        update-types:
```
[the remark this answers is not in the document]
Comment (djm204) [reply .github/dependabot.yml]: Addressed in bfc5156b: switched the guard to parse Dependabot YAML with js-yaml so every group and inline list form is recognized, require @franken/* exclusions on every npm group (including dependency-type groups), and require the @franken/* ignore rule without update-types filters so security updates are blocked too. Added regression coverage for these cases.

File: scripts/check-dependabot-supply-chain.mjs [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+    if (!ignoresInternalScope) {
+      failures.push(
+        `npm Dependabot entry ${directory} must ignore ${INTERNAL_SCOPE_PATTERN} for major/minor/patch updates; internal workspace releases are not registry-driven.`,
+      );
+    }
+
+    for (const [name, group] of Object.entries(update.groups ?? {})) {
+      if (!group || typeof group !== 'object') continue;
+      if (!hasPattern(group.patterns, '*')) continue;
```
[the remark this answers is not in the document]
Comment (djm204) [reply scripts/check-dependabot-supply-chain.mjs]: Addressed in bfc5156b: switched the guard to parse Dependabot YAML with js-yaml so every group and inline list form is recognized, require @franken/* exclusions on every npm group (including dependency-type groups), and require the @franken/* ignore rule without update-types filters so security updates are blocked too. Added regression coverage for these cases.

**Comment by djm204** [comment]
@codex review

File: scripts/check-dependabot-supply-chain.mjs [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+  }
+
+  for (const update of npmUpdates) {
+    const directory =
+      typeof update.directory === 'string' ? update.directory : '<missing directory>';
+    const ignoreRules = asArray(update.ignore);
+    const ignoresInternalScope = ignoreRules.some((rule) => {
+      if (!rule || typeof rule !== 'object') return false;
+      const updateTypes = asArray(rule['update-types']);
+      return rule['dependency-name'] === INTERNAL_SCOPE_PATTERN && updateTypes.length === 0;
```
[the remark this answers is not in the document]
Comment (djm204) [reply scripts/check-dependabot-supply-chain.mjs]: Addressed in ce4a8990: the guard now rejects filtered @franken/* ignore rules (including versions filters), rejects npm Dependabot entries that set target-branch because security updates are evaluated on the default branch, and adds regression tests for both bypass cases.

[the remark this answers is not in the document]
Comment (djm204) [reply scripts/check-dependabot-supply-chain.mjs]: Addressed in ce4a8990: the guard now rejects filtered @franken/* ignore rules (including versions filters), rejects npm Dependabot entries that set target-branch because security updates are evaluated on the default branch, and adds regression tests for both bypass cases.

**Comment by djm204** [comment]
@codex review
